File: linkedList/singly.py
# Our methods are:
# Count, index, insert, remove, append, pop, print

class Node:
    def __init__(self, data):
        self.data = data
        self.ptr = None

class linkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def append(self, data):
        nodeToAppend = Node(data)
        if not self.head:
            self.head = nodeToAppend
            self.tail = nodeToAppend
        else:
            self.tail.ptr = nodeToAppend
            self.tail = nodeToAppend
        self.length += 1
    
    def print(self):
        ptrToIncrement = self.head
        if not ptrToIncrement:
            print("Linked list doesn't have any elements")
            return
        while ptrToIncrement:
            print(ptrToIncrement.data, end =" ")
            ptrToIncrement = ptrToIncrement.ptr
        print("")

    def pop(self):
        if not self.head:
            print("Linked list doesn't have any elements")
            return None
        if self.head is self.tail:
            valToReturn = self.head.data
            self.head = None
            self.tail = None
            return valToReturn
        ptrToIncrement = self.head
        valToReturn = self.tail.data
        while (ptrToIncrement.ptr is not self.tail):
            ptrToIncrement = ptrToIncrement.ptr
        self.tail = ptrToIncrement
        ptrToIncrement.ptr = None
        self.length -= 1
        return valToReturn

    def count(self, data):
        counter = 0
        ptrToIncrement = self.head
        while ptrToIncrement:
            if ptrToIncrement.data == data:
                counter += 1
            ptrToIncrement = ptrToIncrement.ptr
        return counter

    def index(self, data):
        i = 0
        ptrToIncrement = self.head
        while ptrToIncrement:
            if ptrToIncrement.data == data:
                return i
            i += 1
            ptrToIncrement = ptrToIncrement.ptr
        return -1

    def insert(self, index, data):
        if index < 0: # Checks if index is valid
            print("Index must be 0 or positive.")
            return
        if not self.head: # Checks if the linked list is empty
            self.append(data)
            return
        if index == 0: # Case1: Insert at head
            nodeToAdd = Node(data)
            nodeToAdd.ptr = self.head
            self.head = nodeToAdd
            self.length += 1
        elif index < self.length: # Case2: Insert in the middle
            if self.head == self.tail: #Checks if the linked list has 1 element, if that is the case, append the element.
                self.append(data)
            else: # If the linked list is larger than one element, perform the insertion
                i = 1
                ptrToIncrement = self.head
                while i < index:
                    ptrToIncrement = ptrToIncrement.ptr
                    i += 1
                ptrSecond = ptrToIncrement.ptr
                nodeToAdd = Node(data)
                ptrToIncrement.ptr = nodeToAdd
                nodeToAdd.ptr = ptrSecond
                self.length += 1
        else: # If the index is bigger than the current length, append
            self.append(data)

test = linkedList()
for i in range(5):
    test.append(i)

File: linkedList/singly.c
#include <stdio.h>
#include <stdlib.h>
#define LEN(linkedListPtr) (linkedListPtr)->length
#define HEAD(linkedListPtr) (linkedListPtr)->head
#define TAIL(linkedListPtr) (linkedListPtr)->tail
#define DATA(NodePtr) (NodePtr)->data
#define NEXT(NodePtr) (NodePtr)->next

/* Our methods are
 * print, append, count, index, insert, remove, pop, delete */

/* Here are the structs used */

/* Struct for a node */
typedef struct Node
{
	int data;
	struct Node *next;
} Node;

/* Struct for the linked list */
typedef struct linkedList
{
	Node *head;
	Node *tail;
	int length;
}linkedList;

linkedList *createLinkedList(int);
void printLinkedList(linkedList *);
void append(linkedList *, int);
Node *createNode(int);
void pop(linkedList *);

Node *createNode(int data)
{
	Node *Node = malloc(sizeof(Node));
	Node->data = data;
	Node->next = NULL;
	return Node;
}

void printLinkedList(linkedList *linkedList)
{
	int i, len = LEN(linkedList);
	Node *currentNode = HEAD(linkedList);
	for (i = 0; i < len; i++)
	{
		printf("%d ", DATA(currentNode));
		currentNode = NEXT(currentNode);
	}
	printf("\n");
}

linkedList *createLinkedList(int data)
{
	Node *Node = createNode(data);
	linkedList *linkedList = malloc(sizeof(linkedList));
	HEAD(linkedList) = Node;
	TAIL(linkedList) = Node;
	LEN(linkedList) = 1;
	return linkedList;
}

void append(linkedList *linkedList, int data)
{
	Node *Node;
	if (LEN(linkedList) == 0)
	{
		linkedList = createLinkedList(data);
		return;
	}
	Node = createNode(data);
	NEXT(TAIL(linkedList)) = Node;
	TAIL(linkedList) = Node;
	LEN(linkedList)++;
}

void pop(linkedList *linkedList)
{
	
}

int main()
{
	linkedList *linkedList;
	linkedList = createLinkedList(10);
	printLinkedList(linkedList);
	return 0;
}
